rocprim/include/rocprim/device/detail/device_adjacent_find.hpp

```diff
@@ -0,0 +1,156 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#ifndef ROCPRIM_DEVICE_DETAIL_DEVICE_ADJACENT_FIND_HPP_
+#define ROCPRIM_DEVICE_DETAIL_DEVICE_ADJACENT_FIND_HPP_
+
+#include "device_config_helper.hpp"
+#include "ordered_block_id.hpp"
+
+#include "../../block/block_load.hpp"
+#include "../../block/block_reduce.hpp"
+#include "../../intrinsics/thread.hpp"
+
+BEGIN_ROCPRIM_NAMESPACE
+
+namespace detail
+{
+namespace adjacent_find
+{
+template<class OutputT, class IdT>
+ROCPRIM_KERNEL __launch_bounds__(1)
+void init_adjacent_find(OutputT*              reduce_output,
+                        ordered_block_id<IdT> ordered_tile_id,
+                        const size_t          size)
+{
+    // Reset output value.
+    *reduce_output = size;
+
+    // Reset ordered_block_id.
+    ordered_tile_id.reset();
+}
+
+template<typename Config,
+         typename TransformedInputIterator,
+         typename ReduceIndexIterator,
+         typename BinaryPred,
+         typename OrderedTileIdType>
+ROCPRIM_KERNEL
+#ifndef DOXYGEN_DOCUMENTATION_BUILD
+__launch_bounds__(device_params<Config>().kernel_config.block_size)
+#endif
+void block_reduce_kernel(TransformedInputIterator transformed_input,
+                         ReduceIndexIterator      reduce_output,
+                         const std::size_t        size,
+                         BinaryPred               op,
+                         OrderedTileIdType        ordered_tile_id)
+{
+    static constexpr adjacent_find_config_params params     = device_params<Config>();
+    static constexpr unsigned int                block_size = params.kernel_config.block_size;
+    static constexpr unsigned int items_per_thread          = params.kernel_config.items_per_thread;
+    static constexpr unsigned int items_per_tile            = block_size * items_per_thread;
+
+    using transformed_input_type =
+        typename std::iterator_traits<TransformedInputIterator>::value_type;
+    using block_reduce_type = ::rocprim::block_reduce<
+        transformed_input_type,
+        block_size,
+        block_reduce_algorithm::raking_reduce>; // TODO?: params.block_reduce_method>;
```

Comment:
Yes, it can be removed. It was considered in the beginning to tune for the block reduce method but later discarded (couldn't see relevant differences in performance).